lib/mlir/Dialect/Clift/IR/CliftParser.h

```diff
@@ -0,0 +1,201 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <map>
+
+#include "llvm/ADT/SmallVector.h"
+
+#include "mlir/IR/OpImplementation.h"
+
+#include "revng/Support/Assert.h"
+
+#include "revng-c/mlir/Dialect/Clift/IR/CliftAttributes.h"
+
+namespace mlir::clift {
+
+class AsmRecursionGuardBase {
+  using Map = std::map<uint64_t, void *>;
+
+  std::pair<Map::iterator, bool> Insert;
+
+public:
+  explicit AsmRecursionGuardBase(const uint64_t ID);
+
+  AsmRecursionGuardBase(const AsmRecursionGuardBase &) = delete;
+  AsmRecursionGuardBase &operator=(const AsmRecursionGuardBase &) = delete;
+
+  ~AsmRecursionGuardBase();
+
+protected:
+  void *get() const { return Insert.first->second; }
+
+  void set(void *const Value) {
+    revng_assert(Insert.second);
+    revng_assert(Insert.first->second == nullptr);
+    Insert.first->second = Value;
+  }
+};
+
+template<typename ObjectT>
+class AsmRecursionGuard : AsmRecursionGuardBase {
+public:
+  using AsmRecursionGuardBase::AsmRecursionGuardBase;
+
+  [[nodiscard]] ObjectT get() const {
+    return ObjectT(static_cast<
+                   typename ObjectT::ImplType *>(AsmRecursionGuardBase::get()));
+  }
+
+  void set(ObjectT Value) { AsmRecursionGuardBase::set(Value.getImpl()); }
+};
```

Comment:
While I understand the desire to make a better Guard, I think the result is more confusing than the previous version with `getVisitedType` and `onTypeVisited`.

Now we have this wrappers with more layers of abstraction, with an API `get`/`set` that has non-informative naming.
Honestly I find harder to convince myself that the current version of the client code is bug-free than with the previous version. It feels a little bit over-engineered for solving a specific local problem.

So, pick one of the following:
1. revert to `getVisitedType` and `onTypeVisited`
2. improve and document the new guard
  - why `Asm` in the name?
  - what is the general problem it tries to solve? with the previous names it was kinda easier to guess, now it isn't
  - what are the contracts of constructors/destructors?
  - what are the contracts of `get`/`set`
  - do we really need the `Base` version? on the one hand you add a template parameter, but then you use `void *` under the hood :thinking:
  
Honestly I'd just stick with 1.

Reply:
> why Asm in the name?

Because it relates to the parsing and printing of MLIR assembly. Same as `mlir::AsmPrinter` etc.

> what is the general problem it tries to solve?

The main point is to use RAII for safety. With this, the user cannot forget to erase the key from the map, even in the case of early returns.

> do we really need the `Base` version?

That allows me to define the thread local out-of-line, which is something I do as a rule. The codegen for inline thread locals is atrocious.

I can add comments explaining these things, but I much prefer 2 over 1 just for the RAII. As an alternative, if we have a scope guard utility, I could use that as well. But I really want to do the clean up in a destructor.

Reply:
Apparently we don't have a generic scope guard, so if we want to go that route, one would have to be added. That would go in `revng` I suppose.

Reply:
Oh, I totally forgot that LLVM has one which was already being used here.

Reply:
> Oh, I totally forgot that LLVM has one which was already being used here.

Great!

Reply:
It's the `Review changes (to be squashed)` commit. The link gets outdated when I make changes.

Reply:
Ah lovely!